python_modules/dagster/dagster/_core/definitions/decorators/asset_decorator.py

```diff
@@ -1187,47 +1203,47 @@ def build_asset_outs(asset_outs: Mapping[str, AssetOut]) -> Mapping[AssetKey, Tu
+
+    Converts the deps to `AssetDep`s
     """
     if non_argument_deps is not None and deps is not None:
         raise DagsterInvalidDefinitionError(
             "Cannot specify both deps and non_argument_deps to @asset. Use only deps instead."
         )
 
-    upstream_asset_deps: Optional[
-        Iterable[Union[CoercibleToAssetKey, AssetsDefinition, SourceAsset]]
-    ] = None
-    if deps is not None:
-        for dep in deps:
-            if isinstance(dep, AssetsDefinition):
-                # Only AssetsDefinition with a single asset can be passed
-                if len(dep.keys) > 1:
-                    raise DagsterInvalidDefinitionError(
-                        "Cannot pass a multi_asset AssetsDefinition as an argument to deps."
-                        " Instead, specify dependencies on the assets created by the multi_asset"
-                        f" via AssetKeys or strings. For the multi_asset {dep.node_def.name}, the"
-                        f" available keys are: {dep.keys}."
-                    )
-            elif isinstance(dep, SourceAsset):
-                # no additional type checking needed for SourceAssets
-                continue
-            else:
-                # confirm that dep is coercible to AssetKey
-                try:
-                    AssetKey.from_coercible(dep)
-                except check.CheckError:
-                    raise DagsterInvalidDefinitionError(
-                        f"Cannot pass an instance of type {type(dep)} to deps parameter of @asset."
-                        " Instead, pass AssetsDefinitions or AssetKeys."
-                    )
+    upstream_asset_deps: Optional[Iterable[CoercibleToAssetDep]] = None
+
+    with disable_dagster_warnings():
+        if deps is not None:
+            for dep in deps:
+                if isinstance(dep, AssetsDefinition):
+                    # Only AssetsDefinition with a single asset can be passed
+                    if len(dep.keys) > 1:
+                        raise DagsterInvalidDefinitionError(
+                            "Cannot pass a multi_asset AssetsDefinition as an argument to deps."
+                            " Instead, specify dependencies on the assets created by the"
+                            " multi_asset via AssetKeys or strings. For the multi_asset"
+                            f" {dep.node_def.name}, the available keys are: {dep.keys}."
+                        )
+                else:
+                    # confirm that dep is coercible to AssetDep
+                    try:
+                        AssetDep.from_coercible(dep)
+                    except check.CheckError:
+                        raise DagsterInvalidDefinitionError(
+                            f"Cannot pass an instance of type {type(dep)} to deps parameter of"
+                            " @asset. Instead, pass AssetDeps, AssetsDefinitions, SourceAssets, or"
+                            " AssetKeys."
+                        )
```

Comment:
nesting is pretty rough here. suggest some sort of refactor like extracting into function with early returns

Reply:
this gets massively cleaned up once we add support for multi_assets [here](https://github.com/dagster-io/dagster/pull/16556)